File: app/src/main/java/com/example/pam_project/features/categories/create/CreateCategoryActivity.java
package com.example.pam_project.features.categories.create;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;

import com.example.pam_project.R;
import com.example.pam_project.di.ApplicationContainer;
import com.example.pam_project.di.ApplicationContainerLocator;
import com.example.pam_project.repositories.categories.CategoriesRepository;
import com.example.pam_project.utils.constants.AppColor;
import com.example.pam_project.utils.schedulers.SchedulerProvider;
import com.example.pam_project.utils.validators.FormValidator;
import com.thebluealliance.spectrum.SpectrumPalette;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class CreateCategoryActivity extends AppCompatActivity implements CreateCategoryView {
    private static final AppColor DEFAULT_COLOR = AppColor.BLUE;
    private CreateCategoryPresenter presenter;
    private int selectedColor = DEFAULT_COLOR.getARGBValue();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_category);
        createPresenter();
        Objects.requireNonNull(getSupportActionBar()).setTitle(R.string.activity_title_create_category);
        setUpView();
    }

    private void createPresenter() {
        final ApplicationContainer container = ApplicationContainerLocator.locateComponent(this);
        final SchedulerProvider schedulerProvider = container.getSchedulerProvider();
        final CategoriesRepository repository = container.getCategoriesRepository();
        presenter = new CreateCategoryPresenter(schedulerProvider, repository);
    }

    private void setUpView() {
        SpectrumPalette palette = findViewById(R.id.create_category_palette_color);
        final int[] colors = new int[AppColor.values().length];
        for (int i = 0; i < colors.length; i++) {
            colors[i] = AppColor.values()[i].getARGBValue();
        }
        palette.setColors(colors);
        palette.setSelectedColor(selectedColor);
        palette.setOnColorSelectedListener(color -> selectedColor = color);
    }

    @Override
    public void onStart() {
        super.onStart();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.check_action_bar, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        long itemId = item.getItemId();
        final EditText categoryNameInput = findViewById(R.id.create_category_name_input);

        if (itemId == R.id.check_add_button) {
            final String categoryName = categoryNameInput.getText().toString();
            final AppColor color = AppColor.fromARGBValue(selectedColor);
            boolean validForm = FormValidator.validate(getApplicationContext(), createInputMap(categoryName, categoryNameInput));

            String colorName = DEFAULT_COLOR.name();
            if (color != null) {
                colorName = color.name();
            }

            if (validForm) {
                presenter.insertCategory(categoryName, colorName);
                finish();
            }
        }

        return super.onOptionsItemSelected(item);
    }

    private Map<EditText, String> createInputMap(String categoryName, EditText categoryNameInput) {
        Map<EditText, String> map = new HashMap<>();
        map.put(categoryNameInput, categoryName);
        return map;
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.onViewDetached();
    }
}

File: app/src/main/java/com/example/pam_project/features/categories/create/CreateCategoryActivity.kt
package com.example.pam_project.features.categories.create

import android.os.Bundle
import android.view.*
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.pam_project.R
import com.example.pam_project.di.ApplicationContainerLocator
import com.example.pam_project.utils.constants.AppColor
import com.example.pam_project.utils.validators.FormValidator
import com.thebluealliance.spectrum.SpectrumPalette
import java.util.*

class CreateCategoryActivity : AppCompatActivity(), CreateCategoryView {
    private lateinit var presenter: CreateCategoryPresenter
    private var selectedColor = DEFAULT_COLOR.aRGBValue

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_category)
        createPresenter()
        supportActionBar?.setTitle(R.string.activity_title_create_category)
        setUpView()
    }

    private fun createPresenter() {
        val container = ApplicationContainerLocator.locateComponent(this)
        val schedulerProvider = container?.schedulerProvider
        val repository = container?.categoriesRepository
        presenter = CreateCategoryPresenter(schedulerProvider, repository, this)
    }

    private fun setUpView() {
        val palette = findViewById<SpectrumPalette>(R.id.create_category_palette_color)
        val colors = IntArray(AppColor.values().size)
        for (i in colors.indices) {
            colors[i] = AppColor.values()[i].aRGBValue
        }
        palette.setColors(colors)
        palette.setSelectedColor(selectedColor)
        palette.setOnColorSelectedListener { color: Int -> selectedColor = color }
    }

    public override fun onStart() {
        super.onStart()
    }

    override fun onCategoryInsertedError() {
        Toast.makeText(applicationContext, getString(R.string.error_category_create), Toast.LENGTH_LONG).show()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.check_action_bar, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId.toLong()
        val categoryNameInput = findViewById<EditText>(R.id.create_category_name_input)
        if (itemId == R.id.check_add_button.toLong()) {
            val categoryName = categoryNameInput.text.toString()
            val color: AppColor? = AppColor.fromARGBValue(selectedColor)
            val validForm = FormValidator.validate(applicationContext, createInputMap(categoryName, categoryNameInput))
            var colorName = DEFAULT_COLOR.name
            if (color != null) {
                colorName = color.name
            }
            if (validForm) {
                presenter.insertCategory(categoryName, colorName)
                finish()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun createInputMap(categoryName: String, categoryNameInput: EditText): Map<EditText, String?> {
        val map: MutableMap<EditText, String?> = HashMap()
        map[categoryNameInput] = categoryName
        return map
    }

    public override fun onStop() {
        super.onStop()
        presenter.onViewDetached()
    }

    companion object {
        private val DEFAULT_COLOR = AppColor.BLUE
    }
}